libopae++/include/opaec++/log.h

```diff
@@ -0,0 +1,215 @@
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+#include <memory>
+#include <sstream>
+
+namespace opae {
+namespace fpga {
+namespace internal {
+
+/**
+ * @brief wrapped_stream wraps the logging functionality in libopae-c.
+ * It is what is returned from a logger object when it
+ * begins a stream sequence. Its destructor is called when the sequence ends.
+ */
+class wrapped_stream
+{
+ public:
+
+ /**
+  * @brief create a wrapped_stream object with a given level
+  *
+  * @param level The log level expected by the logger in libopae-c
+  * @param new_line Used to determine whether or not a newline is
+  *                 automatically appended to the stream
+  */
+  wrapped_stream(int level, bool new_line = false);
+
+  /**
+   * @brief copy constructor of wrapped_stream that copies internal
+   *        stream data structures
+   *
+   * @param other The other wrapped_stream object to copy from
+   */
+  wrapped_stream(const wrapped_stream & other);
+
+  /**
+   * @brief assignment operator of a wrapped_stream that copies internal
+   *        stream data structures
+   *
+   * @param other The other wrapped_stream object to copy from
+   *
+   * @return A reference to `this` after copying internal data
+   */
+  wrapped_stream & operator=(const wrapped_stream & other);
+
+  /**
+   * @brief wrapped_stream destructor which indicates the end of the stream
+   * sequence
+   */
+  ~wrapped_stream();
+
+  /**
+   * @brief Templated stream operator
+   *
+   * @tparam T The type of the variable being streamed
+   * @param v The variable to stream
+   *
+   * @return A reference to `this`
+   */
+  template<typename T>
+  wrapped_stream& operator<<(const T& v)
+  {
+      sstream_ << v;
+      return *this;
+  }
+
+  /**
+   * @brief Overload of the stream operator used for stream manipulators
+   *
+   * @param manip The stream manipulator
+   *
+   * @return A reference to `this`
+   */
+  wrapped_stream& operator<<(std::ostream& (*manip)(std::ostream&));
+ private:
+  std::stringstream sstream_;
+  int level_;
+  char fmt_[4];
+};
+
+/**
+ * @brief Named logger object used to format log messages with the process
+ * id, the logger name, and the level of the log message
+ */
+class logger {
+ public:
+  /**
+   * @brief log level enumeration that roughly maps to the log levels in
+   * libopae-c
+   */
+  enum class level : int
+  {
+      none        = -1,
+      error       = 0,
```

Comment:
What's the significance of the level values, ie why not just -1..N ?

Reply:
This was pretty much carried over from the first log.cpp in the other c++ libraries (except the order is now reversed). When I did that one, I modeled it after the Python logging library which skips numbers in between their logging levels which I think is done to allow others to define custom log levels. 
I defined these so that you can get to the level in `libopae-c` by dividing by 20.
This is how these levels map to the levels in the `libopae-c`:

C++ enum |C++ level   | C level   | C enum                     
-------------|------------|----------|-----------------------
error           |   0             |         0    | FPGA_LOG_ERROR
exception   | 10             |         0    | FPGA_LOG_ERROR 
fatal            | 15             |         0    | FPGA_LOG_ERROR 
wan            | 20             |         1    | FPGA_LOG_MESSAGE
info             | 25             |         1    | FPGA_LOG_MESSAGE
debug         | 40             |         2    | FPGA_LOG_DEBUG